Feature/147 housing available create endpoint

VitaRox opened the pull request:
I've tested several cases using Swagger docs and Adminer.

# Test Cases 
* existing Pittsburgh, PA address from app.pittsburghhousing.org/properties;
* novel Pittsburgh address based on a real one from app.pittsburghhousing.org/properties;
* a novel Pittsburgh address from Zillow;
* a real Pittsburgh address from Zillow;

-- screenshots attached to show results;
<img width="1003" alt="existingPAAddressFromAppPittsburghhousing" src="https://user-images.githubusercontent.com/24485205/88998844-cd107280-d2a7-11ea-8f7a-1b5d3de98feb.png">
<img width="971" alt="existingPAAddressFromZillow" src="https://user-images.githubusercontent.com/24485205/88998847-d0a3f980-d2a7-11ea-9681-122e7878a14b.png">

<img width="987" alt="novelPAAddressBasedOnARealOneFromZillow" src="https://user-images.githubusercontent.com/24485205/88998861-d8fc3480-d2a7-11ea-984f-31542d1c71c2.png">

<img width="1276" alt="multipleTestsAdminer" src="https://user-images.githubusercontent.com/24485205/88998889-e74a5080-d2a7-11ea-9427-de147f033ac7.png">

-Vita Harvey 30.July.2020

VitaRox committed:
- Initial commit;
- Begins to add in housing-available.routes;

VitaRox committed:
- Fleshes out .routes file; unsure about import requirements;
- Adds imports to .controllers, returns empty object currently;

VitaRox committed:
- Initial commit:
- contains some copy-pasted code stolen from property.controllers.js;
- This branch: fixes error in it's own naming convention (missing "feature/...");

VitaRox committed:
- Replaces instances of propertyObject with housingAvailableObject;
- changes 'var' to 'let', line #53; best practice;
- Thanks for letting me work on this project! @author github.com/VitaRox;

VitaRox committed:
- Removes unnecessary code from property.controllers.js;
- Adds in several lines assigning values to our HousingAvailObject to housing-available.controllers.js;
- Adds in an 'auth' property to 'config' in housing-available.routes.js;
- It's a beautiful day in Seattle; go get some sunshine if it's pretty where you are;

VitaRox committed:
- Adds in api schema defintion to housing-avail.models.js;
- Adds 'createHousingAvail' helper function in housing-avail.controllers.js;

VitaRox committed:
- Comments out "details {}" in api (housing-avail.models) and in housing-avail.controllers;
- Adds a bathroomsValidIncrement;
- Slightly rescopes our createHousingAvailable function so it works;
- Basically runs; there may be an issue with address validation now;

VitaRox committed:
- Tested and works with Pittsburgh addresses, both ones which have already been stored in our db and novel addresses;
- Matches LocationID data to addresses once it finds them;
- I rewrote ~ lines 220-225 in housingAvail.controllers to reflect "Boolean Zen";  I am feeling at one with the command line interface;
- Adds an ensureNonnegativeValue() function to make sure user can't enter a value < 0 for either bedrooms OR bathrooms;
- TODO: ensure that bathroomsValidIncrement() works as expected with values that are multiples of .25 for bathrooms, as well as fail cases for decimal-value number of bathrooms that are not multiples of .25;

zaclittleberry commented on lib/account/account.routes.js:
```diff
@@ -29,6 +29,7 @@ module.exports = {
         method: 'POST',
         path: '/accounts',
         handler: controllers.postAccount,
+        // TODO: "options"?
```
no need to add a comment here

zaclittleberry commented on lib/landlord/landlord.controllers.js:
```diff
@@ -4,9 +4,7 @@ module.exports = (models) => {
   const Sequelize = require('sequelize');
   const Op = Sequelize.Op;
   const landlordModels = require('./landlord.models');
-  const accountLib = require('../account/account.controllers')(models).lib;
```
removing this is probably what is breaking tests. please revert changes

VitaRox replied:
This is so weird.  I don't think I made any changes to this, but maybe I accidentally removed it when thought I was looking at my controller.  I'll fix it.

zaclittleberry commented on lib/property/property.controllers.js:
```diff
@@ -677,9 +677,7 @@ module.exports = (models) => {
     if (remainder === 0) {
       return true;
     }
-
```
while the spaces here aren't needed, it's preferred not to make changes to files that weren't part of your ticket. please revert

VitaRox replied:
Fair enough.  I forgot I did that haha.  Tbh, I really wanna rewrite it to just say "return (remainder === 0);"  I'm an adherent of boolean zen, personally speaking lol.

zaclittleberry commented on package-lock.json:
```diff
@@ -31,11 +31,18 @@
       "dev": true
     },
     "@hapi/boom": {
-      "version": "7.4.2",
```
changes shouldn't be needed to package-lock. please revert

zaclittleberry commented on package.json:
```diff
@@ -15,6 +15,7 @@
   "author": "Zac Littleberry",
   "license": "ISC",
   "dependencies": {
+    "@hapi/boom": "^9.1.0",
```
please revert.

zaclittleberry reviewed (changes requested):
there's a number of files that you should revert changes on, and I think that will fix your tests.

VitaRox committed:
- Fixes diff issues probably caused by my updating some stuff absentmindedly and/or my fanatical adherence to boolean zen;
- These fixes occurred in package.json and package-lock.json in housingdb_api;

VitaRox committed:
- Fixes a couple of small diffs that I missed when reverting/fixing things;

VitaRox committed:
- Fixes one little extra whitespace char that was somehow added;

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+    })
+    .catch(error => {
+      return error;
+    });
```
helper functions shouldn't have their own.then or catch, because the error won't bubble up. please remove this section.

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+        return h
+          .response(housingAvailableInstance);
+      } catch (e) {
+        return e;
```
we never want to just return an error. Some errors are useful to a user, but some aren't and shouldn't make their way back. If our create fails I think we want to throw a Boom.badImplementation. and you can return the h.response(housingAvailableInstance); after the try..catch block.

zaclittleberry commented on lib/housing-available/housing-available.models.js:
```diff
@@ -44,9 +36,32 @@ module.exports = {
+    title: Joi.string(),
+    body: Joi.string(),
+    address: Joi.string(),
+    bedrooms: Joi.number().integer(),
```
please set the .min and .max on bedrooms like we do in property.models.js: .api

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+  }
+
+  // Use to ensure user doesn't enter number of bedrooms or bathrooms less than 0;
+  function ensureNonnegativeValue(value) {
```
Joi validation should already be ensuring values are non-negative. No need to do it again.

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+      let bedrooms = Number(request.payload.bedrooms);
+      if (!ensureNonnegativeValue(bedrooms)) {
+        throw Boom.badData("Bedrooms cannot be a negative number.")
+      }
```
Shouldn't be a need to do this. see below on function.

zaclittleberry commented on lib/housing-available/housing-available.models.js:
```diff
@@ -44,9 +36,32 @@ module.exports = {
+
+  id: Joi.object().keys({
+    id: Joi.string().guid()
+  }),
+  // Define api;
+  api: Joi.object().keys({
```
I was able to submit the payload "{}"; something in this payload should be required. I think at least contact, title, and body would be an acceptable start.

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+        housingAvailableObject.contact = request.payload.contact;
+      }
+
+      if (request.payload.metadata) {
+        housingAvailableObject.metadata = request.payload.metadata;
+      }
```
metadata isn't on the payload. and users can't/shouldn't be able to submit it (it's usually something calculated by the system to help the system display info without having to calculate things on the fly). please remove.

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,229 @@
+      }
+
+      // This sets the AuthorId associated with this HA listing to === AuthorId of the post itself;
+      if (request.payload.AuthorId) {
```
there should be an else here that sets AuthorId to account.id if AuthorId isn't in the payload (because it's not required). In the case of HousingAvailable, we never want the content to be un-owned.

zaclittleberry reviewed (changes requested):
some changes needed. but should be pretty straight forward. let me know if you have any questions

VitaRox committed:
- Implements changes requested for pull request feature/147-housing-available-create-endpoint, 7.August.2020;
- Checked using Swagger and Adminer, verified working as before;
- This removes unnecessary helper function, provides a way of associating an account.id with AuthorId if none is otherwise provided,and other improvements;

zaclittleberry commented on lib/housing-available/housing-available.controllers.js:
```diff
@@ -1 +1,209 @@
+      if (request.payload.AuthorId) {
+        housingAvailableObject.AuthorId = request.payload.AuthorId;
+      } else {
+        housingAvailableObject.AuthorId = request.payload.account.id;
```
There's no account object on request.payload. The account I was referring to was the one you initialized on line 20 and then filled in the few lines after that, and then ensured it had a result after that.

so the right side of that expression just needs to be account.id;

zaclittleberry reviewed (changes requested):
Very good! Only 1 change left

VitaRox committed:
- Fixes line 72 in housing-available.controllers.js by assigning to just 'account.id' on the else condition rather
	than 'request.payload.account.id';
